main.rb

```diff
@@ -0,0 +1,107 @@
+module Enumerable
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    i = 0
+    array = to_a
+    while i < array.length
+      yield(array[i])
+      i += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    i = 0
+    while i < length
+      yield(self[i], i)
+      i += 1
+    end
+    self
+  end
```

Comment:
- [ ] when a block is given when self is a range it should return the range itself after calling the given block once for each element in self

```Ruby
  range.my_each_with_index(&block) # returns undefined local variable or method `length' for 5..50:Range
```

- [ ] when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self

```Ruby
  enum.my_each_with_index(&block) # returns undefined local variable or method `length' for 5..50:Range
```

- [ ] when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self

```Ruby
  block = proc { |num, idx| my_each_with_index_output += "Num: #{num}, idx: #{idx}\n" }
  my_each_with_index(&block)
  
  # expected: "Num: [:a, 1], idx: 0\nNum: [:b, 2], idx: 1\nNum: [:c, 3], idx: 2\nNum: [:d, 4], idx: 3\nNum: [:e, 5], idx: 4\n"
  # got: "Num: , idx: 0\nNum: , idx: 1\nNum: , idx: 2\nNum: , idx: 3\nNum: , idx: 4\n"

```

Reply:
We have tried to fix it. I think the problem was there was no defined method for `Ranges`